File: project-{topic}-optimized/src/controllers/task.controller.js
const taskService = require('../services/task.service');
const projectService = require('../services/project.service'); // To check project existence and authorization
const logger = require('../utils/logger');
const httpStatus = require('http-status');
const { USER_ROLES } = require('../config/constants');

/**
 * Check if the user has access to the project
 * @param {string} projectId
 * @param {object} user
 * @returns {Promise<boolean>}
 */
const checkProjectAccess = async (projectId, user) => {
  const project = await projectService.getProjectById(projectId);
  if (!project) return false;

  // Admin/Manager always have access
  if (user.role === USER_ROLES.ADMIN || user.role === USER_ROLES.MANAGER) return true;

  // Members can access if they are the project owner or assigned to a task within it (though not explicitly checked here,
  // we assume project ownership is the primary gate for project-level access for members).
  // For simplicity, a member can only access projects they own. More complex logic might involve team members.
  return project.ownerId === user.id;
};

/**
 * Create a new task
 * POST /api/v1/tasks
 */
const createTask = async (req, res, next) => {
  try {
    const { projectId, assignedToId } = req.body;

    // Validate project access
    const hasAccess = await checkProjectAccess(projectId, req.user);
    if (!hasAccess) {
      return res.status(httpStatus.FORBIDDEN).json({ message: 'You do not have permission to create tasks in this project.' });
    }

    // Optional: Validate assignedToId exists if provided
    if (assignedToId) {
      const assignedUser = await prisma.user.findUnique({ where: { id: assignedToId } });
      if (!assignedUser) {
        return res.status(httpStatus.BAD_REQUEST).json({ message: 'Assigned user not found.' });
      }
    }

    const task = await taskService.createTask(req.body, req.user.id);
    res.status(httpStatus.CREATED).json({
      message: 'Task created successfully',
      task
    });
  } catch (error) {
    logger.error(`Error creating task: ${error.message}`);
    next(error);
  }
};

/**
 * Get all tasks
 * GET /api/v1/tasks
 * Supports filters: projectId, assignedToId, creatorId, status
 */
const getTasks = async (req, res, next) => {
  try {
    const filters = req.query;

    // If user is a MEMBER, only allow them to see tasks they are assigned to or created,
    // or tasks within projects they own. For simplicity, we limit to assigned/created.
    if (req.user.role === USER_ROLES.MEMBER) {
      // Members can only query tasks related to themselves
      filters.assignedToId = filters.assignedToId || req.user.id;
      filters.creatorId = filters.creatorId || req.user.id;
      // If a projectId filter is present, ensure the member has access to that project.
      // This is a basic check; a more robust solution would involve checking all tasks found against project access.
      if (filters.projectId) {
        const hasAccessToProject = await checkProjectAccess(filters.projectId, req.user);
        if (!hasAccessToProject) {
          return res.status(httpStatus.FORBIDDEN).json({ message: 'You do not have permission to view tasks in this project.' });
        }
      }
    }

    const tasks = await taskService.getAllTasks(filters);

    // Filter tasks if a member requested and we fetched broadly due to other filters
    const authorizedTasks = tasks.filter(task => {
        if (req.user.role === USER_ROLES.MEMBER) {
            return task.assignedToId === req.user.id || task.creatorId === req.user.id;
        }
        return true;
    });

    res.status(httpStatus.OK).json(authorizedTasks);
  } catch (error) {
    logger.error(`Error getting tasks: ${error.message}`);
    next(error);
  }
};

/**
 * Get a task by ID
 * GET /api/v1/tasks/:id
 */
const getTask = async (req, res, next) => {
  try {
    const { id } = req.params;
    const task = await taskService.getTaskById(id);

    if (!task) {
      return res.status(httpStatus.NOT_FOUND).json({ message: 'Task not found' });
    }

    // Authorization: Only assigned user, creator, manager, or admin can view a task
    if (
      task.assignedToId !== req.user.id &&
      task.creatorId !== req.user.id &&
      req.user.role !== USER_ROLES.MANAGER &&
      req.user.role !== USER_ROLES.ADMIN
    ) {
      return res.status(httpStatus.FORBIDDEN).json({ message: 'You do not have permission to view this task.' });
    }

    res.status(httpStatus.OK).json(task);
  } catch (error) {
    logger.error(`Error getting task ${req.params.id}: ${error.message}`);
    next(error);
  }
};

/**
 * Update a task
 * PUT /api/v1/tasks/:id
 */
const updateTask = async (req, res, next) => {
  try {
    const { id } = req.params;
    const existingTask = await taskService.getTaskById(id);

    if (!existingTask) {
      return res.status(httpStatus.NOT_FOUND).json({ message: 'Task not found' });
    }

    // Authorization: Only assigned user, creator, manager, or admin can update a task
    // Non-admins cannot change 'creatorId' or 'projectId'
    if (
      existingTask.assignedToId !== req.user.id &&
      existingTask.creatorId !== req.user.id &&
      req.user.role !== USER_ROLES.MANAGER &&
      req.user.role !== USER_ROLES.ADMIN
    ) {
      return res.status(httpStatus.FORBIDDEN).json({ message: 'You do not have permission to update this task.' });
    }

    // Prevent non-admins/managers from changing creator or project
    if (req.body.creatorId || req.body.projectId) {
      if (req.user.role !== USER_ROLES.ADMIN && req.user.role !== USER_ROLES.MANAGER) {
        return res.status(httpStatus.FORBIDDEN).json({ message: 'Only managers or administrators can change task creator or project.' });
      }
    }

    const updatedTask = await taskService.updateTask(id, req.body);
    res.status(httpStatus.OK).json({ message: 'Task updated successfully', task: updatedTask });
  } catch (error) {
    logger.error(`Error updating task ${req.params.id}: ${error.message}`);
    next(error);
  }
};

/**
 * Delete a task
 * DELETE /api/v1/tasks/:id
 */
const deleteTask = async (req, res, next) => {
  try {
    const { id } = req.params;
    const existingTask = await taskService.getTaskById(id);

    if (!existingTask) {
      return res.status(httpStatus.NOT_FOUND).json({ message: 'Task not found' });
    }

    // Authorization: Only creator, manager, or admin can delete a task
    if (
      existingTask.creatorId !== req.user.id &&
      req.user.role !== USER_ROLES.MANAGER &&
      req.user.role !== USER_ROLES.ADMIN
    ) {
      return res.status(httpStatus.FORBIDDEN).json({ message: 'You do not have permission to delete this task.' });
    }

    await taskService.deleteTask(id);
    res.status(httpStatus.NO_CONTENT).send(); // 204 No Content for successful deletion
  } catch (error) {
    logger.error(`Error deleting task ${req.params.id}: ${error.message}`);
    next(error);
  }
};

module.exports = {
  createTask,
  getTasks,
  getTask,
  updateTask,
  deleteTask
};

File: project-{topic}-advanced/src/controllers/TaskController.cpp
```cpp
#include "TaskController.hpp"

TaskController::TaskController(std::shared_ptr<TaskService> task_service,
                               std::shared_ptr<ProjectService> project_service,
                               std::shared_ptr<UserService> user_service)
    : task_service_(task_service), project_service_(project_service), user_service_(user_service) {}

// Helper function for authorization logic (can be moved to a more general auth helper)
bool TaskController::canManageProjectTasks(const AuthContext& ctx, int project_id) {
    if (AuthMiddleware::hasRole(ctx, UserRole::ADMIN)) {
        return true; // Admin can manage any project's tasks
    }
    // Check if the authenticated user is the owner of the project
    return project_service_->isProjectOwner(project_id, ctx.user_id);
}


void TaskController::registerRoutes(crow::App<crow::CORSHandler, AuthMiddleware, RateLimitMiddleware>& app) {
    // Get all tasks (Admin only, or all tasks for owned projects?)
    // For simplicity, let's allow any authenticated user to see all tasks for now, but restrict to admin or project owner for CRUD.
    CROW_ROUTE(app, "/api/v1/tasks")
        .methods(crow::HTTPMethod::GET)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx) {
        return try_catch_handler([&]() -> crow::response {
            // For now, allow any authenticated user to view all tasks.
            // In a real app, you'd filter by projects they are involved in.
            std::vector<Task> tasks = task_service_->getAllTasks();
            return crow::response(200, JSONConverter::toJSON(tasks).dump());
        });
    });
    
    // Get tasks by project ID
    CROW_ROUTE(app, "/api/v1/projects/<int>/tasks")
        .methods(crow::HTTPMethod::GET)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx, int project_id) {
        return try_catch_handler([&]() -> crow::response {
            // Check if project exists
            if (!project_service_->getProjectById(project_id).has_value()) {
                throw NotFoundException("Project with ID " + std::to_string(project_id) + " not found.");
            }
            std::vector<Task> tasks = task_service_->getTasksByProjectId(project_id);
            return crow::response(200, JSONConverter::toJSON(tasks).dump());
        });
    });

    // Create a new task for a specific project
    CROW_ROUTE(app, "/api/v1/projects/<int>/tasks")
        .methods(crow::HTTPMethod::POST)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx, int project_id) {
        return try_catch_handler([&]() -> crow::response {
            // Authorization: Only project owner or admin can create tasks in a project
            if (!canManageProjectTasks(ctx, project_id)) {
                throw ForbiddenException("Access denied: Not authorized to create tasks in this project.");
            }

            nlohmann::json req_body = JSONConverter::parse(req.body);
            TaskCreateDTO create_dto = TaskCreateDTO::fromJson(req_body);
            create_dto.project_id = project_id; // Set project_id from URL parameter

            // If assigned_user_id is provided, validate it exists
            if (create_dto.assigned_user_id.has_value() && !user_service_->getUserById(create_dto.assigned_user_id.value()).has_value()) {
                throw BadRequestException("Assigned user with ID " + std::to_string(create_dto.assigned_user_id.value()) + " does not exist.");
            }

            Task new_task = task_service_->createTask(create_dto);
            return crow::response(201, JSONConverter::toJSON(new_task).dump());
        });
    });

    // Get task by ID
    CROW_ROUTE(app, "/api/v1/tasks/<int>")
        .methods(crow::HTTPMethod::GET)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx, int task_id) {
        return try_catch_handler([&]() -> crow::response {
            std::optional<Task> task = task_service_->getTaskById(task_id);
            if (!task.has_value()) {
                throw NotFoundException("Task with ID " + std::to_string(task_id) + " not found.");
            }
            // Authorization: Any authenticated user can view tasks. For stricter access,
            // check if user is owner of task's project or assigned to the task.
            return crow::response(200, JSONConverter::toJSON(task.value()).dump());
        });
    });

    // Update task by ID
    CROW_ROUTE(app, "/api/v1/tasks/<int>")
        .methods(crow::HTTPMethod::PUT)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx, int task_id) {
        return try_catch_handler([&]() -> crow::response {
            std::optional<Task> existing_task_opt = task_service_->getTaskById(task_id);
            if (!existing_task_opt.has_value()) {
                throw NotFoundException("Task with ID " + std::to_string(task_id) + " not found.");
            }
            
            // Authorization: Only project owner or admin can update tasks
            if (!canManageProjectTasks(ctx, existing_task_opt->project_id)) {
                throw ForbiddenException("Access denied: Not authorized to update tasks in this project.");
            }

            nlohmann::json req_body = JSONConverter::parse(req.body);
            TaskUpdateDTO update_dto = TaskUpdateDTO::fromJson(req_body);

            // If assigned_user_id is updated, validate it exists
            if (update_dto.assigned_user_id.has_value() && update_dto.assigned_user_id.value() != 0) {
                if (!user_service_->getUserById(update_dto.assigned_user_id.value()).has_value()) {
                    throw BadRequestException("Assigned user with ID " + std::to_string(update_dto.assigned_user_id.value()) + " does not exist.");
                }
            }
            // If project_id is updated, ensure the user has rights to the new project as well
            if (update_dto.project_id.has_value() && update_dto.project_id.value() != existing_task_opt->project_id) {
                if (!canManageProjectTasks(ctx, update_dto.project_id.value())) {
                     throw ForbiddenException("Access denied: Not authorized to move task to the target project.");
                }
            }


            Task updated_task = task_service_->updateTask(task_id, update_dto);
            return crow::response(200, JSONConverter::toJSON(updated_task).dump());
        });
    });

    // Delete task by ID
    CROW_ROUTE(app, "/api/v1/tasks/<int>")
        .methods(crow::HTTPMethod::DELETE)
        (app.get_middleware<AuthMiddleware>(), [&](const crow::request& req, AuthContext& ctx, int task_id) {
        return try_catch_handler([&]() -> crow::response {
            std::optional<Task> existing_task_opt = task_service_->getTaskById(task_id);
            if (!existing_task_opt.has_value()) {
                throw NotFoundException("Task with ID " + std::to_string(task_id) + " not found.");
            }

            // Authorization: Only project owner or admin can delete tasks
            if (!canManageProjectTasks(ctx, existing_task_opt->project_id)) {
                throw ForbiddenException("Access denied: Not authorized to delete tasks in this project.");
            }

            task_service_->deleteTask(task_id);
            return crow::response(204); // No content
        });
    });
}
```